### ISSUE TITLE
[HOLD for payment 2023-06-29] [$1000] Delete popup is stuck and screen blinks for few seconds when user tries to delete a message and user gets removed from workspace at that instance

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17846]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app on any device and login with user A
2. Open the app on any other device and login with user B (ensure that user A is present in one of the workspace to which user B is admin)
3. On user A device, open workspace to which user B is admin
4. On user A device, right click or long press on any of the self message and click on delete message to open delete confirmation popup
5. From user B device, remove user A from that workspace
6. Observe on user A device, when user B removes user A from workspace, chat blinks by showing not access page, chat page and again not access page and also the delete popup is still open

## Expected Result: 
App should also close all the popups open on user A device when user B remove user A from workspace. App should also not blink the #announce chat for user A when user B removes user A from workspace
## Actual Result: 
App keeps the delete popup open even though user A is no longer able to able that chat. App also blinks the chat few times before displaying 'You don't have access' screen

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/233814365-a544b62f-e828-4ca2-afa5-fe2d01c44548.mp4

https://user-images.githubusercontent.com/43996225/233814372-d03315b4-61d8-4f05-b278-1bb3f9463f98.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1682176575452809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019ed03865c0010a59</li>
        <li>Upwork Job ID: 1651681139754299392</li>
        <li>Last Price Increase: 2023-06-09</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by alexpensify** [comment]
I'll try to test soon.

**Comment by MelvinBot** [comment]
@alexpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by alexpensify** [comment]
This one is still on my radar to test.

**Comment by alexpensify** [comment]
I'm able to replicate 

![2023-04-26_15-08-22](https://user-images.githubusercontent.com/8107070/234714025-9a47a4e6-375f-4a4e-bcd9-dd85699187d1.jpg)

![2023-04-26_15-08-40](https://user-images.githubusercontent.com/8107070/234714042-ceb874d9-4449-4743-9c29-ff0119afd40a.jpg)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @bondydaa (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by alexpensify** [comment]
@bondydaa - I think we take this to the room to identify how it should function but let me know if you have ideas then we assign it to External. Thanks!

**Comment by bondydaa** [comment]
Hmm I don't see why `engineering` label was added here, this should be fine for a contributor to work on so adding external label. Me asking in the room for ideas is the same as posting the `external` label no?

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~019ed03865c0010a59

**Comment by MelvinBot** [comment]
Current assignee @alexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Current assignee @bondydaa is eligible for the External assigner, not assigning anyone new.

**Comment by manish-tapadar-18** [comment]
@bondydaa If you are using websocket or any real time mechanism then easily you can solve this problem.
When someone remove a user from workspace we can easily send a real time packet and close the modal or popup.

**Comment by MelvinBot** [comment]
📣 @manish-tapadar-18! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by bondydaa** [comment]
Yep we are using web-socket via a service called Pusher which is how the report is able to get updated and show "not available" anymore. I think we just need to also dismiss the modal component somehow which I didn't dig into. 

If you haven't already checkout https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md for how we typically expect proposals to come in (https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job)

**Comment by chafikchaban** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
ActionContext menu delete modal is not closed when user gets removed from workspace.
( same goes for the context menu which IMO should also be closed in this case )
### What is the root cause of that problem?
the `ReportActionContextMenu` is not aware of the changes, thus it keeps its states ( open **context menu** or **delete modal** )
### What changes do you think we should make in order to solve the problem?
when a report update is received, in the case of a `null` report, invoke `hideDeleteModal` and `hideContextMenu` from `ReportActionContextMenu` in [Report.js#L688](https://github.com/Expensify/App/blob/32fbb2c44aeea0e112f8eb0ca9ea4911e160f499/src/libs/actions/Report.js#L688)

_**results:**_

https://user-images.githubusercontent.com/30408070/235005953-003766c6-27bb-4a11-bc24-d4154363055d.mov


https://user-images.githubusercontent.com/30408070/235005975-e4efb24c-9308-4c89-95af-419129e4c97c.mov



### What alternative solutions did you explore? (Optional)
no better solution i can think of

**Comment by MelvinBot** [comment]
📣 @chafikchaban! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/231263719-a8a3b7d3-36a0-4d7f-8ad4-cebe1f560fa4.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by chafikchaban** [comment]
Contributor details
Your Expensify account email: chafikchaaben@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01568ab50c600ff939

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Delete popup is stuck and screen blinks for few seconds when user tries to delete a message and user gets removed from workspace at that instance

### What is the root cause of that problem?

PopoverReportActionContextMenu is a global component, it will not be automatically hidden when the ReportActionItem is deleted

### What changes do you think we should make in order to solve the problem?

When I tried to switch the Report by the browser's back button, I found that the [`PopoverReportActionContextMenu.PopoverWithMeasuredContent`, `PopoverReportActionContextMenu.ConfirmModal`] was not hidden either

So I think we should `show` and `hide ReportActionContextMenu` at the `same place`, so I think we can judge `this.state.isContextMenuActive` in `ReportActionItem.componentWillUnmount`, and execute `ReportActionContextMenu.hideContextMenu` and `ReportActionContextMenu.hideDeleteModal`

**Comment by bondydaa** [comment]
Thanks for the proposals! 

@chafikchaban I think with your proposal the context menu / delete modal would clear any time a report was deleted, not just the report I was trying to edit, right? Like if I have reportID 1 open, go to delete a message then I'm removed from reportID 2 by a workspace admin even though reportID 1 was open for me once onyx got the update for the reportID 2 we'd call to close the context menu?

@hellohublot I think your proposal is on the right track, though we are in the process of updating all of our components to use react hooks so we wouldn't want to add a new lifecycle method. 

The PR to convert `ReportActionItem` is here https://github.com/Expensify/App/pull/16635/, I'm wondering if maybe we hold this issue until that is merged to avoid a potentially nasty conflict, cc @jasperhuangg @marcaaron thoughts?

@0xmiroslav what do you think?

**Comment by bondydaa** [comment]
bump @0xmiroslav @jasperhuangg @marcaaron thoughts on whether we should hold this until the `ReportActionItem` refactor is done https://github.com/Expensify/App/pull/16635? Not that it's going to fix this bug but since we will most likely need to tap into some of the new hooks it's probably better to wait til that is done.

**Comment by jasperhuangg** [comment]
@bondydaa Thanks for the bump. I was holding off on completing the refactor since it wasn't a high enough priority item, but it sounds to me like this is cause to get it done now. I'll push it forward today

**Comment by marcaaron** [comment]
Seems like we are pretty close to merging that other one so let's get it over the finish line then wrap this up!

**Comment by bondydaa** [comment]
okay going to throw this on hold then until that PR is merged then.

**Comment by hellohublot** [comment]
@bondydaa @0xmiroslav 
Thanks, I guessed my [proposal](https://github.com/Expensify/App/issues/17846#issuecomment-1526974818) was able to solve the delete box `stucks`, so I continued to investigate how to solve screen `blinks`

After deleting a person from a room, the interface display sequence:
`ServerOnyx.update({ report_xxx: null })` -> `FullPageNotFoundView` -> `ReportActionsSkeletonView` -> `FullPageNotFoundView`

So I tried to add `if (routeReportID === prevProps.report.reportID && !onyxReportID) { return }` in `ReportScreen.componentDidUpdate`, but this will not work, because we ignore not only when it is deleted by the server, but also when using `deeplink`, The `onyxReportID` will be set to empty for a short time

so I think the following video may not have a perfect solution. If it can be solved, it must have been solved long ago, or we can separate it into a `separate issue`, because it is not related to `delete box` already


https://user-images.githubusercontent.com/20135674/235612710-a89640ba-9bc4-4618-be17-9d872c341a3b.mov

**Comment by alexpensify** [comment]
Is there any feedback for @0xmiroslav following their latest tests? Thanks!

**Comment by alexpensify** [comment]
Bumping this one again since we need proposal feedback. Thanks!

**Comment by alexpensify** [comment]
Whoops, sorry I missed that this is on hold but thank you for the update.

**Comment by alexpensify** [comment]
The PR we are waiting on is still under review.

**Comment by alexpensify** [comment]
Still on hold

**Comment by alexpensify** [comment]
Still on hold due to this PR - https://github.com/Expensify/App/pull/18809

@bondydaa this appears to be taking longer. Should we move this one to a weekly?

**Comment by alexpensify** [comment]
@dylanexpensify - This one is on hold, assigning a BZ team member to update it when the HOLD is over. Thanks!

**Comment by bondydaa** [comment]
ya sure 👍

**Comment by dylanexpensify** [comment]
Any update here, @bondydaa? 🙇‍♂️

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂  Thanks @bfitzexpensify!!

**Comment by bfitzexpensify** [comment]
https://github.com/Expensify/App/pull/18809 has been deployed to staging — I'll message again once it hits prod, and then @0xmiroslav you can review @hellohublot's proposal

**Comment by alexpensify** [comment]
@bfitzexpensify - I appreciate your help here! I'm back online and removing your assignment.

**Comment by alexpensify** [comment]
https://github.com/Expensify/App/pull/18809 hit production! @bondydaa are we clear to remove the hold?

**Comment by bondydaa** [comment]
okay off hold.

the ReportActionItem component is now using hooks and has the `useEffect` hook already https://github.com/Expensify/App/blob/77c9d55879be3d4fb86118af6a2909881a05fe3d/src/pages/home/report/ReportActionItem.js#L112-L121

so my hunch is we might need to tap into that if possible here but I would suggest any previous proposals take a look at the updated component before resubmitting https://github.com/Expensify/App/pull/18809/

**Comment by alexpensify** [comment]
Waiting for proposals

**Comment by alexpensify** [comment]
Waiting for proposals

**Comment by alexpensify** [comment]
If no movement soon, then I'll shop around for help here since this one is ready for a proposal.

**Comment by alexpensify** [comment]
Still waiting for proposals

**Comment by alexpensify** [comment]
No update

**Comment by alexpensify** [comment]
@0xmiroslav - when you get a chance, can you please review the recent proposal? Thanks!

**Comment by alexpensify** [comment]
@bondydaa - I think you are working limited time tomorrow. Let me know if I need to find someone else to help review @0xmiroslav's feedback. Thanks!

**Comment by alexpensify** [comment]
Easy Melvin, this one is waiting for a review. We should be back on track this week.

**Comment by alexpensify** [comment]
@arosiclair - I've assigned the Engineering GH because Bondy is working limited time. We need feedback here: 

https://github.com/Expensify/App/issues/17846#issuecomment-1579249756

Thank you, I'd like to get this one moving forward.

**Comment by arosiclair** [comment]
@hellohublot can you expand on this a bit:

> So I think we should show and hide ReportActionContextMenu at the same place, so I think we can judge ReportActionContextMenu.isActiveReportAction(this.props.action.reportActionID) in ReportActionItem.useEffect(() => { return () => {  } }, []), and execute ReportActionContextMenu.hideContextMenu and ReportActionContextMenu.hideDeleteModal

How specifically would you use `useEffect` in `ReportActionItem`? Write pseudocode if necessary

**Comment by arosiclair** [comment]
Hmm this effect doesn't have any dependencies so it won't run when the report action is deleted. Also we probably don't need to check if the context menu or delete modal are already active, we can just attempt to dismiss them. If they aren't active, that should just result in a no-op. 

Can you update your proposal to address this and test to make sure it works in practice? @0xmiroslav can you work with him to iron out this proposal?

**Comment by arosiclair** [comment]
Oh okay cool I didn't see that we were returning a cleanup function. Thanks for the testing LGTM. Let's implement that proposal. Be sure to include a short comment about why we need to hide these on unmount.

**Comment by alexpensify** [comment]
PR is under review -- 👏🏼

**Comment by alexpensify** [comment]
Awesome! Now, we are waiting for the automation to kick in here.

**Comment by alexpensify** [comment]
@hellohublot, and @0xmiroslav please apply at https://www.upwork.com/jobs/~019ed03865c0010a59 and I can prepare for the pay period on 6/29. Thank you!

**Comment by dhanashree-sawant** [comment]
Hi @alexpensify, should I also apply for reporting bonus?

**Comment by alexpensify** [comment]
@dhanashree-sawant - yes please!

**Comment by dhanashree-sawant** [comment]
Thanks, I have applied.

**Comment by alexpensify** [comment]
Tomorrow is the payment date. @dhanashree-sawant, @hellohublot, and @0xmiroslav please apply at https://www.upwork.com/jobs/~01522d777e8a7b8f13 Thank you!

**Comment by dhanashree-sawant** [comment]
Thanks @alexpensify , if possible, can you send me invite to the job?

**Comment by alexpensify** [comment]
@dhanashree-sawant, @hellohublot, and @0xmiroslav - the proposals have been prepared in Upwork. Please accept and I can complete the payment. Thank you!

**Comment by dhanashree-sawant** [comment]
Thanks a lot! I have accepted the offer.

**Comment by alexpensify** [comment]
Alright, @dhanashree-sawant, @hellohublot, and @0xmiroslav have been paid via Upwork. I've closed the job too, great work here!
